python/hartree_fock/impurity.py

```diff
@@ -127,53 +183,177 @@ def f(Sigma_HF_flat):
                         mpi.report('Sigma_HF[\'%s\']:' % name)
                         mpi.report(bl)
+
                 for bl, G0_bl in self.G0_iw:
                     self.G_iw[bl] << inverse(inverse(G0_bl) - self.Sigma_HF[bl])
+               
+                G_dens = self.G_iw.density()
+                with np.printoptions(suppress=True, precision=4):
+                  for name, bl in G_dens.items():
+                    mpi.report('Final G_dens[\'%s\']:' % name)
+                    mpi.report(bl)
 
             else:
                 mpi.report('Hartree-Fock solver did not converge successfully.')
                 mpi.report(root_finder['message'])
+                self.Sigma_HF = unflatten(root_finder['x'], self.gf_struct, self.force_real)
+                mpi.report('Calculated self energy:')
+                with np.printoptions(suppress=True, precision=4):
+                    for name, bl in self.Sigma_HF.items():
+                        mpi.report('Sigma_HF[\'%s\']:' % name)
+                        mpi.report(bl)
+
+                for bl, G0_bl in self.G0_iw:
+                    self.G_iw[bl] << inverse(inverse(G0_bl) - self.Sigma_HF[bl])
+               
+                G_dens = self.G_iw.density()
+                with np.printoptions(suppress=True, precision=4):
+                  for name, bl in G_dens.items():
+                    mpi.report('Final G_dens[\'%s\']:' % name)
+                    mpi.report(bl)
+
 
     def interaction_energy(self):
         """ Calculate the interaction energy
 
         """
         E = 0
         for bl, gbl in self.G_iw:
-            E += 0.5 * np.trace(self.Sigma_HF[bl].dot(gbl.density()))
+            E += 0.5 * np.trace(self.Sigma_int[bl].dot(gbl.density().real))
+        return E
+    
+    def DC_energy(self):
+        """ Calculate the DC energy
+
+        """
+        E = 0
+        for bl, gbl in self.G_iw:
+            E += 0.5 * np.trace(self.Sigma_DC[bl].dot(gbl.density().real))
         return E
+    
+    def _reinitialize_sigma(self, Sigma_guess):
+        """ Changes in place the sigma with the average over frequencies of a given Gf object.
+            Used to update the initial guess, must be called before the solve() method
+
+        Parameters
+        ----------
+            Sigma_guess : GfImFreq or GfReFreq object
+        """
+        # super ugly, needs changing
+        for bl in self.Sigma_HF.keys():
+            if self.force_real:
+                self.Sigma_HF[bl] = np.mean(Sigma_guess[bl].data, axis=0).real
+            else:
+                self.Sigma_HF[bl] = np.mean(Sigma_guess[bl].data, axis=0)
+
+        with np.printoptions(suppress=True, precision=4):
+          for name, bl in self.Sigma_HF.items():
+              mpi.report('HARTREE SOLVER: Updated guess for Sigma_HF[\'%s\']:' % name)
+              mpi.report(bl)
+
+
+    
+    def _interface_to_solid_dmft(self, general_params, advanced_params, icrsh):
```

Comment:
I don't really like having solid_dmft specific things here. Can't we move this somehow to solid_dmft, and initialize the solver explicitly? We can discuss this if you think we need it here.

Reply:
Moved to solid_dmft